src/fd.cpp

```diff
@@ -0,0 +1,147 @@
+#ifdef _WIN32
+#include <winsock2.h>
+#endif
+#include <Rcpp.h>
+#include <unistd.h>
+#include <cstdlib>
+#include "later.h"
+#include "callback_registry_table.h"
+
+extern CallbackRegistryTable callbackRegistryTable;
+
+typedef struct ThreadArgs_s {
+  SEXP callback;
+  R_xlen_t num_fds;
+  std::unique_ptr<std::vector<int>> fds;
+  fd_set read_fds;
+  struct timeval tv;
+  int max_fd;
+  int loop;
+  bool flag;
+} ThreadArgs;
+
+static void later_callback(void *arg) {
+
+  std::unique_ptr<std::shared_ptr<ThreadArgs>> argsptr(static_cast<std::shared_ptr<ThreadArgs>*>(arg));
+  std::shared_ptr<ThreadArgs> args = *argsptr;
+
+  SEXP results, call;
+  // TODO: actually pre-allocate call on main thread, values can be safely updated from background thread
+  // only call Rf_eval() here to minimise runtime in later callback.
+  PROTECT(results = Rf_allocVector(LGLSXP, args->num_fds));
+  std::memcpy((void *) DATAPTR_RO(results), args->fds->data(), args->num_fds * sizeof(int));
+  PROTECT(call = Rf_lcons(args->callback, Rf_cons(results, R_NilValue)));
+  Rf_eval(call, R_GlobalEnv);
+  UNPROTECT(2);
+  R_ReleaseObject(args->callback);
+
+}
+
+// CONSIDER: upgrade select() to poll() / WSAPoll().
+// CONSIDER: add method for HANDLES on Windows. Assuming we only accept integer
+// values for both, we could use heuristics: check if it is a valid SOCKET or
+// else assume HANDLE - but this is not fool-proof.
+// Otherwise would require an interface for user to specify type.
+static void *select_thread(void *arg) {
+
+  std::unique_ptr<std::shared_ptr<ThreadArgs>> argsptr(static_cast<std::shared_ptr<ThreadArgs>*>(arg));
```

Comment:
Are you pretty confident in the smart pointer logic? I don't feel confident code reviewing that aspect for you, but if you would like a second set of eyes on this, I can ask someone from the RStudio IDE team to take a look.

Reply:
Confident that it works, and also doesn't seem to leak memory (running valgrind). The code's probably safe enough even using manual memory management if we wanted to push performance, but I thought it best to try and tie in with the rest of the package. I'm no way an expert though so happy to accept comments here.

Reply:
Good enough for me, thanks!